esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -2490,32 +2490,48 @@ bool CFileSprayEx::onCopy(IEspContext &context, IEspCopy &req, IEspCopyResponse
         StringBuffer fileMask;
         constructFileMask(destTitle.str(), fileMask);
 
+        Owned<IUserDescriptor> udesc=createUserDescriptor();
         const char* srcDali = req.getSourceDali();
-        bool supercopy = req.getSuperCopy();
-        if (supercopy)
+        const char* srcu = req.getSrcusername();
+        if (!isEmptyString(srcDali) && !isEmptyString(srcu))
+        {
+            udesc->set(srcu, req.getSrcpassword());
+        }
+        else
         {
             StringBuffer user, passwd;
             context.getUserID(user);
             context.getPassword(passwd);
-            StringBuffer u(user);
-            StringBuffer p(passwd);
-            Owned<INode> foreigndali;
-            if (srcDali)
-            {
-                SocketEndpoint ep(srcDali);
-                foreigndali.setown(createINode(ep));
-                const char* srcu = req.getSrcusername();
-                if(srcu && *srcu)
-                {
-                    u.clear().append(srcu);
-                    p.clear().append(req.getSrcpassword());
-                }
-            }
-            Owned<IUserDescriptor> udesc=createUserDescriptor();
-            udesc->set(u.str(),p.str());
-            if (!queryDistributedFileDirectory().isSuperFile(srcname,udesc,foreigndali))
+            udesc->set(user, passwd);
+        }
+
+        Owned<IDistributedFile> file;
+        if (isEmptyString(srcDali) ||
+            ((strlen(srcname) > 9) && (strnicmp(srcname, "foreign::", 9) == 0)) ||
+            ((strlen(srcname) > 10) && (strnicmp(srcname, "~foreign::", 10) == 0)))
+            file.setown(queryDistributedFileDirectory().lookup(srcname, udesc.get()));
+        else
+        {
+            StringBuffer fileToLookup;
+            fileToLookup.set("foreign::").append(srcDali).append("::");
+            if ((strlen(srcname) > 3) && (strnicmp(srcname, ".::", 3) == 0))
+                fileToLookup.append(srcname + 3);
+            else
+                fileToLookup.append(srcname);
```

Comment:
You should use CDfsLogicalFileName to manipulate (add foreign) and get the correct format.

Reply:
See example code that uses CDfsLogicalFileName::setForeign()

Reply:
I thought that there might be some existing code to do that. I just do not know where. I will try the CDfsLogicalFileName ...